File: sdks/javascript/src/utils/http-client.js
/**
 * HTTP client utility for making API requests using native fetch.
 */

import {
  NetworkError,
  ServerError,
  AuthenticationError,
  RateLimitError,
} from '../exceptions/index.js';

export class HttpClient {
  constructor(config = {}) {
    this.config = {
      timeout: 30000,
      ...config,
    };
  }

  /**
   * Make a GET request.
   */
  async get(url, requestConfig = {}) {
    const response = await this.request(url, {
      method: 'GET',
      ...requestConfig,
    });
    return response;
  }

  /**
   * Make a POST request.
   */
  async post(url, data, requestConfig = {}) {
    const response = await this.request(url, {
      method: 'POST',
      body: data ? JSON.stringify(data) : undefined,
      ...requestConfig,
    });
    return response;
  }

  /**
   * Make a PUT request.
   */
  async put(url, data, requestConfig = {}) {
    const response = await this.request(url, {
      method: 'PUT',
      body: data ? JSON.stringify(data) : undefined,
      ...requestConfig,
    });
    return response;
  }

  /**
   * Make a DELETE request.
   */
  async delete(url, requestConfig = {}) {
    const response = await this.request(url, {
      method: 'DELETE',
      ...requestConfig,
    });
    return response;
  }

  /**
   * Make a POST request with FormData (for file uploads).
   * @param {string} url - URL to post to
   * @param {FormData} formData - FormData object containing the file and other fields
   * @param {Object} requestConfig - Additional request configuration
   * @returns {Promise<Object>} Response data
   */
  async postFormData(url, formData, requestConfig = {}) {
    const response = await this.requestFormData(url, formData, requestConfig);
    return response;
  }

  /**
   * Make a generic HTTP request.
   */
  async request(url, options = {}) {
    const fullUrl = url.startsWith('http') ? url : `${this.config.baseURL}${url}`;
    
    const headers = {
      'Content-Type': 'application/json',
      ...this.config.headers,
      ...options.headers,
    };

    if (this.config.apiKey) {
      headers['Authorization'] = `Bearer ${this.config.apiKey}`;
    }

    // const controller = new AbortController();
    // const timeout = setTimeout(() => {
    //   controller.abort();
    // }, options.timeout || this.config.timeout);

    try {
      const response = await fetch(fullUrl, {
        ...options,
        headers,
        // signal: controller.signal,
      });

      // clearTimeout(timeout);

      if (response && !response.ok) {
        throw this.handleError(response);
      }

      if (!response) {
        throw new NetworkError('No response received');
      }

      const contentType = response.headers?.get('content-type');
      if (contentType && contentType.includes('application/json')) {
        return await response.json();
      }

      return await response.text();
    } catch (error) {
      // clearTimeout(timeout);

      // If it's already a VectorizerError (from handleError), re-throw it
      if (error instanceof Error && error.name && error.name.includes('Error') && error.constructor.name !== 'Error') {
        throw error;
      }

      if (error instanceof Error) {
        // if (error.name === 'AbortError') {
        //   throw new TimeoutError('Request timeout');
        // }
        throw new NetworkError(error.message);
      }

      throw new NetworkError('Unknown network error');
    }
  }

  /**
   * Make a FormData HTTP request (for file uploads).
   * Note: Content-Type header is not set to allow the browser to set it with boundary.
   * @param {string} url - URL to post to
   * @param {FormData} formData - FormData object
   * @param {Object} options - Request options
   * @returns {Promise<Object>} Response data
   */
  async requestFormData(url, formData, options = {}) {
    const fullUrl = url.startsWith('http') ? url : `${this.config.baseURL}${url}`;

    // Don't set Content-Type for FormData - let the browser set it with boundary
    const headers = {};

    if (this.config.headers) {
      Object.assign(headers, this.config.headers);
    }

    if (options.headers) {
      Object.assign(headers, options.headers);
    }

    // Remove Content-Type if set - browser needs to set it for multipart/form-data
    delete headers['Content-Type'];

    if (this.config.apiKey) {
      headers['Authorization'] = `Bearer ${this.config.apiKey}`;
    }

    try {
      const response = await fetch(fullUrl, {
        method: 'POST',
        headers,
        body: formData,
      });

      if (response && !response.ok) {
        throw this.handleError(response);
      }

      if (!response) {
        throw new NetworkError('No response received');
      }

      const contentType = response.headers?.get('content-type');
      if (contentType && contentType.includes('application/json')) {
        return await response.json();
      }

      return await response.text();
    } catch (error) {
      // If it's already a VectorizerError (from handleError), re-throw it
      if (error instanceof Error && error.name && error.name.includes('Error') && error.constructor.name !== 'Error') {
        throw error;
      }

      if (error instanceof Error) {
        throw new NetworkError(error.message);
      }

      throw new NetworkError('Unknown network error');
    }
  }

  /**
   * Handle HTTP errors and convert them to appropriate exceptions.
   */
  handleError(response) {
    const message = `HTTP ${response.status}: ${response.statusText}`;

    switch (response.status) {
      case 401:
        return new AuthenticationError(message);
      case 403:
        return new AuthenticationError('Access forbidden');
      case 404:
        return new ServerError('Resource not found');
      case 429:
        return new RateLimitError('Rate limit exceeded');
      case 500:
      case 502:
      case 503:
      case 504:
        return new ServerError(message);
      default:
        return new ServerError(message);
    }
  }
}

File: sdks/python/utils/http_client.py
"""
HTTP client utility for making API requests using aiohttp.
"""

from typing import Optional, Dict, Any
import aiohttp
import asyncio
import logging

try:
    from ..exceptions import (
        NetworkError,
        ServerError,
        AuthenticationError,
        RateLimitError,
    )
except ImportError:
    from exceptions import (
        NetworkError,
        ServerError,
        AuthenticationError,
        RateLimitError,
    )

# Issue #263: cap Retry-After respect at this many seconds so a
# misconfigured server can't pin a client into a half-hour sleep.
_RETRY_AFTER_MAX_SECONDS = 30
# Floor for parsed Retry-After so a `0` or missing header still yields
# a noticeable backoff rather than busy-looping the server.
_RETRY_AFTER_DEFAULT_SECONDS = 1

logger = logging.getLogger(__name__)


class HTTPClient:
    """HTTP transport client."""
    
    def __init__(
        self,
        base_url: str = "http://localhost:15002",
        api_key: Optional[str] = None,
        timeout: int = 30,
        max_retries: int = 3
    ):
        """
        Initialize HTTP client.
        
        Args:
            base_url: Base URL for HTTP API
            api_key: API key for authentication
            timeout: Request timeout in seconds
            max_retries: Maximum number of retry attempts
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout
        self.max_retries = max_retries
        self._session: Optional[aiohttp.ClientSession] = None
    
    async def _ensure_session(self):
        """Ensure aiohttp session is created."""
        if self._session is None or self._session.closed:
            headers = {"Content-Type": "application/json"}
            if self.api_key:
                # JWT shape → `Authorization: Bearer`; anything else
                # (raw API keys from `POST /auth/keys`) → `X-API-Key`.
                # Server middleware won't fall back Bearer → api_key,
                # so routing has to happen client-side.
                parts = self.api_key.split(".")
                if len(parts) == 3 and all(p for p in parts):
                    headers["Authorization"] = f"Bearer {self.api_key}"
                else:
                    headers["X-API-Key"] = self.api_key
            
            timeout_config = aiohttp.ClientTimeout(total=self.timeout)
            self._session = aiohttp.ClientSession(
                headers=headers,
                timeout=timeout_config
            )
    
    async def close(self):
        """Close the HTTP session."""
        if self._session and not self._session.closed:
            await self._session.close()
            self._session = None
    
    async def request(
        self,
        method: str,
        path: str,
        data: Optional[Dict[str, Any]] = None
    ) -> Any:
        """
        Make an HTTP request.

        Honors `Retry-After` on 429 responses (issue #263): the client
        sleeps for the header's value (capped) and retries up to
        ``max_retries`` times. After exhaustion a ``RateLimitError`` is
        raised so callers can surface the back-pressure to the user.

        Args:
            method: HTTP method
            path: API endpoint path
            data: Request data

        Returns:
            Response data
        """
        await self._ensure_session()

        url = f"{self.base_url}{path}"
        attempts_remaining = self.max_retries
        last_429_text: Optional[str] = None

        while True:
            try:
                async with self._session.request(
                    method,
                    url,
                    json=data if data else None,
                ) as response:
                    if response.status == 429:
                        last_429_text = await response.text()
                        if attempts_remaining <= 0:
                            raise RateLimitError(
                                f"HTTP 429 after {self.max_retries} retries: "
                                f"{last_429_text}"
                            )
                        delay = _parse_retry_after(
                            response.headers.get("Retry-After")
                        )
                        logger.info(
                            "Vectorizer 429 — sleeping %.1fs before retry "
                            "(remaining attempts=%d)",
                            delay,
                            attempts_remaining,
                        )
                        attempts_remaining -= 1
                        await asyncio.sleep(delay)
                        continue

                    if response.status >= 400:
                        error_text = await response.text()
                        raise self._handle_error(response.status, error_text)

                    content_type = response.headers.get('Content-Type', '')
                    if 'application/json' in content_type:
                        return await response.json()
                    return await response.text()

            except (ServerError, AuthenticationError, RateLimitError):
                raise
            except aiohttp.ClientError as e:
                raise NetworkError(f"HTTP request failed: {e}")
            except asyncio.TimeoutError:
                raise NetworkError("Request timeout")
            except Exception as e:
                raise NetworkError(f"Unknown error: {e}")
    
    async def get(self, path: str) -> Any:
        """Make a GET request."""
        return await self.request("GET", path)
    
    async def post(self, path: str, data: Optional[Dict[str, Any]] = None) -> Any:
        """Make a POST request."""
        return await self.request("POST", path, data)
    
    async def put(self, path: str, data: Optional[Dict[str, Any]] = None) -> Any:
        """Make a PUT request."""
        return await self.request("PUT", path, data)
    
    async def delete(self, path: str) -> Any:
        """Make a DELETE request."""
        return await self.request("DELETE", path)
    
    def _handle_error(self, status: int, error_text: str) -> Exception:
        """Handle HTTP errors and convert to appropriate exceptions."""
        message = f"HTTP {status}: {error_text}"

        if status == 401:
            return AuthenticationError(message)
        elif status == 403:
            return AuthenticationError("Access forbidden")
        elif status == 404:
            return ServerError("Resource not found")
        elif status == 429:
            # 429 is handled in `request()` via Retry-After; reaching
            # here means the caller bypassed retry handling, so
            # surface a typed RateLimitError instead of a generic 5xx.
            return RateLimitError(message)
        elif status in (500, 502, 503, 504):
            return ServerError(message)
        else:
            return ServerError(message)


def _parse_retry_after(value: Optional[str]) -> float:
    """Parse a ``Retry-After`` header value (seconds form only).

    Returns a sane default + caps an unreasonably large server hint
    so a misconfigured server can't pin a client into a long sleep.
    """
    if not value:
        return _RETRY_AFTER_DEFAULT_SECONDS
    try:
        seconds = float(value.strip())
    except ValueError:
        return _RETRY_AFTER_DEFAULT_SECONDS
    if seconds <= 0:
        return _RETRY_AFTER_DEFAULT_SECONDS
    return min(seconds, _RETRY_AFTER_MAX_SECONDS)

